File: leetcode-tree-traversal-solutions-pro/src/morris_traversal.py
from src.data_structures import TreeNode

class MorrisTraversal:
    """
    Implements Morris Traversal for Inorder and Preorder.
    Morris Traversal is an O(1) space complexity traversal algorithm.
    It works by temporarily modifying the tree structure to create "threaded" links,
    and then restoring the original structure.
    """

    @staticmethod
    def morris_inorder_traversal(root: TreeNode) -> list[int]:
        """
        Performs an Inorder Traversal using Morris Traversal algorithm.
        Time Complexity: O(N), where N is the number of nodes. Each edge is traversed at most twice.
        Space Complexity: O(1) auxiliary space.

        Algorithm:
        1. Initialize `current` node to `root`.
        2. While `current` is not None:
           a. If `current.left` is None:
              This node has no left child, so it is the next node in inorder sequence.
              Add `current.val` to result.
              Move `current` to its right child.
           b. Else (`current.left` exists):
              Find the rightmost node in the left subtree (predecessor of `current`).
              The predecessor is the node that would be visited just before `current` in inorder.
              `predecessor = current.left`
              While `predecessor.right` is not None AND `predecessor.right` is not `current`:
                  `predecessor = predecessor.right`
              
              i. If `predecessor.right` is None (first time visiting `current`'s left subtree):
                 This means we haven't linked `predecessor` back to `current` yet.
                 Create a temporary link: `predecessor.right = current`.
                 Move `current` to its left child.
              ii. If `predecessor.right` is `current` (second time visiting `current`'s left subtree):
                  This means the left subtree has been fully processed, and we're returning to `current`.
                  Remove the temporary link: `predecessor.right = None`.
                  Add `current.val` to result (now `current` is visited after its left subtree).
                  Move `current` to its right child.
        """
        result = []
        current = root

        while current:
            if current.left is None:
                # No left child, so current node is the next in inorder sequence
                result.append(current.val)
                current = current.right # Move to the right subtree
            else:
                # Left child exists, find the inorder predecessor
                predecessor = current.left
                while predecessor.right is not None and predecessor.right != current:
                    predecessor = predecessor.right
                
                if predecessor.right is None:
                    # First time visiting current.left's subtree
                    # Establish a temporary link (thread) from predecessor to current
                    predecessor.right = current
                    current = current.left # Move to the left subtree
                else:
                    # Second time visiting current.left's subtree (left subtree is processed)
                    # This means we've returned via the thread.
                    # Remove the thread and visit current node.
                    predecessor.right = None # Remove the temporary link
                    result.append(current.val)
                    current = current.right # Move to the right subtree
        return result

    @staticmethod
    def morris_preorder_traversal(root: TreeNode) -> list[int]:
        """
        Performs a Preorder Traversal using Morris Traversal algorithm.
        Time Complexity: O(N).
        Space Complexity: O(1) auxiliary space.

        Algorithm (similar to Inorder, but visit Root when it's first encountered):
        1. Initialize `current` node to `root`.
        2. While `current` is not None:
           a. If `current.left` is None:
              Add `current.val` to result (Root is visited).
              Move `current` to its right child.
           b. Else (`current.left` exists):
              Find the rightmost node in the left subtree (predecessor of `current`).
              `predecessor = current.left`
              While `predecessor.right` is not None AND `predecessor.right` is not `current`:
                  `predecessor = predecessor.right`

              i. If `predecessor.right` is None (first time visiting `current`'s left subtree):
                 Add `current.val` to result (Root is visited, as this is preorder).
                 Create a temporary link: `predecessor.right = current`.
                 Move `current` to its left child.
              ii. If `predecessor.right` is `current` (second time visiting `current`'s left subtree):
                  This means the left subtree has been fully processed.
                  Remove the temporary link: `predecessor.right = None`.
                  Move `current` to its right child (no visit here, as it was visited in step i).
        """
        result = []
        current = root

        while current:
            if current.left is None:
                # No left child, so current node is the next in preorder sequence after itself
                result.append(current.val)
                current = current.right # Move to the right subtree
            else:
                # Left child exists, find the inorder predecessor
                predecessor = current.left
                while predecessor.right is not None and predecessor.right != current:
                    predecessor = predecessor.right
                
                if predecessor.right is None:
                    # First time visiting current's left subtree
                    # Visit current node now (Preorder)
                    result.append(current.val)
                    # Establish a temporary link (thread) from predecessor to current
                    predecessor.right = current
                    current = current.left # Move to the left subtree
                else:
                    # Second time visiting current's left subtree (left subtree is processed)
                    # This means we've returned via the thread.
                    # Remove the thread. Current node was already visited.
                    predecessor.right = None # Remove the temporary link
                    current = current.right # Move to the right subtree
        return result

File: coding-interview-tree-traversal-plus/src/problems/morrisTraversal.js
```javascript
const { TreeNode } = require('../utils/treeNode');

/**
 * Problem: Binary Tree Inorder Traversal (Morris Traversal - O(1) Space)
 *
 * Given the root of a binary tree, return the inorder traversal of its nodes' values.
 * This implementation uses Morris Traversal, which achieves O(1) auxiliary space complexity.
 *
 * Morris Traversal works by temporarily modifying the tree structure by creating
 * "threaded" links to simulate stack behavior without using an explicit stack.
 *
 * The algorithm can be summarized as follows:
 * 1. Initialize `current` to `root`.
 * 2. While `current` is not null:
 *    a. If `current` has no left child:
 *       - Add `current.val` to the result.
 *       - Move `current` to `current.right`. (No left subtree to traverse, so current node is visited, then go right)
 *    b. If `current` has a left child:
 *       - Find the rightmost node in `current`'s left subtree (this is the "predecessor").
 *       - The predecessor is found by going left once, then right repeatedly until `right` is null or `right` is `current`.
 *       - If the predecessor's `right` child is `null` (first time visiting this link):
 *         - Make `predecessor.right` point to `current`. (Create a thread)
 *         - Move `current` to `current.left`. (Continue exploring the left subtree)
 *       - If the predecessor's `right` child is `current` (already visited left subtree, returning to root):
 *         - Add `current.val` to the result. (Visit current node)
 *         - Set `predecessor.right` back to `null`. (Break the thread to restore tree)
 *         - Move `current` to `current.right`. (Move to right subtree)
 *
 * Example:
 *     1
 *      \
 *       2
 *      /
 *     3
 * Inorder: [1, 3, 2]
 */

/**
 * Approach: Morris Inorder Traversal
 *
 * Time Complexity: O(N) - Although there are nested loops, each edge in the tree is traversed at most a constant number of times.
 *                          Specifically, each node's right child link is modified at most twice (once to establish the thread,
 *                          once to break it). All other operations are constant time per node.
 * Space Complexity: O(1) - It uses no extra space beyond the output list and a few pointers.
 * @param {TreeNode|null} root - The root of the binary tree.
 * @returns {number[]} An array containing the inorder traversal of node values.
 */
function morrisInorderTraversal(root) {
    const result = [];
    let current = root;

    while (current !== null) {
        if (current.left === null) {
            // Case 1: No left child.
            // This means we have visited all nodes in the left subtree (if any)
            // and processed the current node. So, add its value to the result
            // and move to the right child.
            result.push(current.val);
            current = current.right;
        } else {
            // Case 2: Has a left child.
            // Find the inorder predecessor (rightmost node in the left subtree).
            let predecessor = current.left;
            while (predecessor.right !== null && predecessor.right !== current) {
                predecessor = predecessor.right;
            }

            // Subcase 2.1: Predecessor's right child is null.
            // This is the first time we are visiting `current` from its predecessor.
            // Create a temporary link (thread) from predecessor to current.
            // Then, move `current` to its left child to continue exploring the left subtree.
            if (predecessor.right === null) {
                predecessor.right = current; // Create thread
                current = current.left;
            } else {
                // Subcase 2.2: Predecessor's right child is `current`.
                // This means we have already visited `current`'s left subtree
                // and are returning to `current` via the thread.
                // Now, it's time to visit `current` itself.
                result.push(current.val); // Visit current node
                predecessor.right = null; // Break the thread to restore the tree structure
                current = current.right; // Move to the right child to continue traversal
            }
        }
    }

    return result;
}

module.exports = {
    morrisInorderTraversal
};
```